components/datetime/src/format.rs

```diff
@@ -0,0 +1,56 @@
+use super::date::DateTime;
+// use super::pattern::{parse_pattern, FieldType};
+use crate::fields::{FieldLength, FieldSymbol};
+use crate::pattern::{Pattern, PatternItem};
+use std::fmt;
+
+pub struct FormattedDateTime<'l> {
+    pub(crate) pattern: &'l Pattern<'static>,
+    pub(crate) display_names: &'l crate::provider::DisplayNames<'static>,
+    pub(crate) date_time: &'l DateTime,
+}
+
+impl<'l> fmt::Display for FormattedDateTime<'l> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_pattern(self.pattern, self.display_names, &self.date_time, f)
+    }
+}
+
+fn format_number(
+    result: &mut impl fmt::Write,
+    num: usize,
+    length: &FieldLength,
+) -> Result<(), std::fmt::Error> {
+    write!(result, "{:0>width$}", num, width = (*length as u8) as usize)
+}
+
+pub fn write_pattern(
```

Comment:
This function is the bread and butter.  It's clearly not finished yet, and there isn't enough here right now for me to give meaningful feedback.  I will look at this file again once it is more fleshed out.

Reply:
My intention so far was to explore how it would look like in Rust. It does what it is supposed to do - it takes a `Pattern` and writes a formatted version of it to output based on data stored on `DateTimeFormat` from `DataProvider`.

What foundational pieces are missing?

Reply:
The other fields (eras, day periods, quarters, etc) are not here.  But I think the general direction is good.